### PR TITLE
chore(*): Use new createContentDigest helper

### PR DESCRIPTION
**Pull request by samscha**
**Note**: this is my first pr to Gatsby so if I'm missing something (best practices included), please do tell me about them and I'll update this/future pr's.

Added createContentDigest helper for gatsby-transformer-remark and gatsby-source-filesystem.

I opened this pr right now mainly because I wanted to make sure I was adding the createContentDigest helper to these two plugins correctly and not breaking anything (before going on to other plugins).

If it's more convenient, and I have verification I correctly added the createContentDigest helper to these two plugins, this pr can be closed and I can open a new pr later with more plugins that have the createContentDigest helper added.

Thanks!

from
#8805

### CODE REVIEW

**Comment by samscha** [comment]
**Yes** I can definitely work on re-implementing tests to use `createContentDigest`.

For clarification, this would mean that the snapshots, instead of having `contentDigest`, would have their unique md5 hash?

Thanks!

**Comment by DSchau** [comment]
@samscha that is exactly right!

File: packages/gatsby-source-contentful/src/cache-image.js [hunk 1]
```diff
@@ -1,9 +1,10 @@
-const crypto = require(`crypto`)
 const { resolve, parse } = require(`path`)
 
 const axios = require(`axios`)
 const { pathExists, createWriteStream } = require(`fs-extra`)
 
+const createContentDigest = require(`../../gatsby/src/utils/create-content-digest`)
```
Comment (pieh) [inline packages/gatsby-source-contentful/src/cache-image.js]: we should pass this from caller and not import directly here

Comment (pieh) [reply packages/gatsby-source-contentful/src/cache-image.js]: we should only import it directly for tests

Comment (samscha) [reply packages/gatsby-source-contentful/src/cache-image.js]: thank you for the comment! getting from the caller makes a lot more sense! I fixed all but two of these instances in previous commits.

I'm going to work on removing the import from gatsby-source-contentful and gatsby-transformer-sqip (both use the cache-image code mentioned above)

**Comment by m-allanson** [comment]
Thanks for this PR @samscha 👍 I've fixed up the merge conflicts, this LGTM now. @DSchau or @pieh, do you want to take another look at this?

**Comment by DSchau** [comment]
@m-allanson will take a look; thank you!

**Comment by pieh** [comment]
There will be problems with `gatsby-source-filesystem`/`createRemoteFileNode` - we would need to pass `createContentDigest` there which isn't currently done

**Comment by LekoArts** [comment]
@pieh So would this PR also need to update `createRemoteFileNode`? And plugins that use that + the helper their (peer)Dependency?

**Comment by pieh** [comment]
I would honestly try to leave out `gatsby-source-filesystem` out of the updates here or add fallback to `createRemoteFileNode` if `createContentDigest` isn't passed to it, to not break any 3rd party plugins using this API

**Comment by LekoArts** [comment]
@samscha Hi 👋 I also feel like it would be a good idea to leave `gatsby-source-filesystem` out of this PR. Could you make the requested changes? 😊

File: packages/gatsby-transformer-documentationjs/src/gatsby-node.js [hunk 2]
```diff
@@ -1,10 +1,6 @@
 const documentation = require(`documentation`)
-const crypto = require(`crypto`)
-const digest = str =>
-  crypto
-    .createHash(`md5`)
-    .update(str)
-    .digest(`hex`)
+const createContentDigest = require(`../../gatsby/src/utils/create-content-digest`)
+const digest = str => createContentDigest(str)
```
Comment (muescha) [inline packages/gatsby-transformer-documentationjs/src/gatsby-node.js]: istead of this lambda: why not rename all `digest` to `createContentDigest`

File: packages/gatsby-source-wordpress/src/normalize.js [hunk 3]
```diff
@@ -14,11 +14,7 @@ const restrictedNodeFields = [`id`, `children`, `parent`, `fields`, `internal`]
  *
  * @param {any} str
  */
-const digest = str =>
-  crypto
-    .createHash(`md5`)
-    .update(str)
-    .digest(`hex`)
+const digest = str => createContentDigest(str)
```
Comment (muescha) [inline packages/gatsby-source-wordpress/src/normalize.js]: is it not better to rename all `digest` to `createContentDigest`?

File: packages/gatsby-transformer-sqip/src/generate-sqip.js [hunk 4]
```diff
@@ -30,10 +30,7 @@ module.exports = async function generateSqip(options) {
     mode,
   }
 
-  const optionsHash = crypto
-    .createHash(`md5`)
-    .update(JSON.stringify(sqipOptions))
-    .digest(`hex`)
+  const optionsHash = createContentDigest(JSON.stringify(sqipOptions))
```
Comment (muescha) [inline packages/gatsby-transformer-sqip/src/generate-sqip.js]: i think stringify can be removed here?
(i found it also on more places)

File: packages/gatsby-source-lever/src/normalize.js [hunk 5]
```diff
@@ -36,7 +25,7 @@ async function createGraphQLNode(ent, type, createNode, store, cache) {
   }
   node = recursiveAddFields(ent, node)
   node.internal.content = JSON.stringify(node)
-  node.internal.contentDigest = digest(stringify(node))
+  node.internal.contentDigest = createContentDigest(stringify(node))
```
Comment (muescha) [inline packages/gatsby-source-lever/src/normalize.js]: remove stringify?

File: packages/gatsby-transformer-react-docgen/src/on-node-create.js [hunk 6]
```diff
@@ -1,11 +1,7 @@
-import crypto from "crypto"
+import createContentDigest from "../../gatsby/src/utils/create-content-digest"
 import parseMetadata from "./parse"
 
-const digest = str =>
-  crypto
-    .createHash(`md5`)
-    .update(str)
-    .digest(`hex`)
+const digest = str => createContentDigest(str)
```
Comment (muescha) [inline packages/gatsby-transformer-react-docgen/src/on-node-create.js]: ist not better to rename all `digest` to `createContentDigest`

**Comment by KyleAMathews** [comment]
Thanks!

**Comment by KyleAMathews** [comment]
Hey @samscha! We unfortunately had to revert this — looks like there were a couple of problems that we didn't catch before merging. Could you put up a new PR with these fixes?

Basically we need to keep the old implementations around as a fallback as many people won't upgrade `gatsby` when upgrading/installing a package so the `utils.js` file wouldn't be there. Also, currently that file is blocked from being published so that'd need fixed here https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/package.json#L147-L152

For the fallbacks, perhaps move them to a separate file e.g. `fallbacks.js` and add a TODO at the top to remove them in gatsby v3?

**Comment by samscha** [comment]
Got it. Will open up a new PR!

**Comment by KyleAMathews** [comment]
Thanks!

The `fallback.js` file would be in the `src` directory for each plugin. It would try to require the core code and if that fails, use the fallback code. This way when v3 comes, all we need to do is to remove the fallback.js file and replace any code importing code from it to just use the core code.

**Pull request merged**